src/Materials/babylon.uniformBuffer.ts

```diff
@@ -0,0 +1,584 @@
+module BABYLON {
+
+    // Pool for avoiding memory leaks
+    var MAX_UNIFORM_SIZE = 256;
+    var _tempBuffer = new Float32Array(MAX_UNIFORM_SIZE);
+
+    export class UniformBuffer {
+        private _engine: Engine;
+        private _buffer: WebGLBuffer;
+        private _data: number[];
+        private _bufferData: Float32Array;
+        private _dynamic: boolean;
+        private _uniformName: string;
+        private _uniformLocations: { [key:string]:number; };
+        private _uniformSizes: { [key:string]:number; };
+        private _uniformLocationPointer: number;
+        private _needSync: boolean;
+        private _cache: Float32Array;
+        private _noUbo: boolean;
+        private _currentEffect: Effect;
+
+        /**
+         * Uniform buffer objects.
+         * 
+         * Handles blocks of uniform on the GPU.
+         *
+         * If WebGL 2 is not available, this class falls back on traditionnal setUniformXXX calls.
+         *
+         * For more information, please refer to : 
+         * https://www.khronos.org/opengl/wiki/Uniform_Buffer_Object
+         */
+        constructor(engine: Engine, data?: number[], dynamic?: boolean) {
+            this._engine = engine;
+            this._noUbo = engine.webGLVersion === 1;

```

Comment:
As this will be heavily used here is an idea:
we declare container function like this:
public updateMatrix3x3: (name: string, matrix: Float32Array) => Effect;

And we create two functions:
private _updateMatrix3x3ForUniform(name: string, matrix: Float32Array): Effect {
}
private _updateMatrix3x3ForEffect(name: string, matrix: Float32Array): Effect {
}


then in the constructor based on _noUBO value we affect one of the private function to the public one. So direct call, no test